src/forkserver.rs

```diff
@@ -0,0 +1,310 @@
+    /// Wait for parent to fire next execution
+    pub(crate) const NEXT: ChildRet = 0;
+    /// Current execution done with a crash found
+    pub(crate) const FOUND_CRASH: ChildRet = 1;
+    /// Edge generation event. This is always followed by generated edge PC.
+    ///
+    /// This will never be sent when child finished its execution.
+    pub(crate) const TSL_REQUEST: ChildRet = 2;
+    /// The child process has exited.
+    ///
+    /// This will never be sent from child to parent. Instead, this is a phantom
+    /// state used for forkserver parent state management.
+    pub(crate) const EXITED: ChildRet = 3;
+    /// The child process found a crash, and is the last execution of one persistent loop.
+    ///
+    /// This is used for indicating the parent that it should not expect the child is in
+    /// persistent loop any more.
+    pub(crate) const FOUND_CRASH_AND_EXITED: ChildRet = 4;
+}
+
+type AflChildRet = afl_child_ret::ChildRet;
+
+/// Forkserver parent for UnicornAFL
+pub struct UnicornAflForkserverParent<'a, D, OT, H>
+where
+    D: 'a,
+{
+    /// Executor.
+    ///
+    /// You could drop the parent and take ownership back when parent
+    /// returns from [`start_forkserver`][libafl_targets::start_forkserver], which
+    /// indicates that it is the child process, and parent is useless anymore (the
+    /// owned resources have been transferred to the executor itself).
+    pub(crate) executor: UnicornAflExecutor<'a, D, OT, H>,
+    child_pipe_r: Option<PipeReader>,
+    child_pipe_w: Option<PipeWriter>,
+    parent_pipe_r: Option<PipeReader>,
+    parent_pipe_w: Option<PipeWriter>,
+    last_child_pid: Option<i32>,
+    last_child_ret: AflChildRet,
+    old_sigchld_handler: Option<SigHandler>,
+    wifsignaled: i32,
+}
+
+impl<'a, D, OT, H> UnicornAflForkserverParent<'a, D, OT, H>
+where
+    D: 'a,
+{
+    /// Create a new forkserver parent
+    pub fn new(executor: UnicornAflExecutor<'a, D, OT, H>) -> Self {
+        Self {
+            executor,
+            child_pipe_r: None,
+            child_pipe_w: None,
+            parent_pipe_r: None,
+            parent_pipe_w: None,
+            last_child_pid: None,
+            last_child_ret: afl_child_ret::EXITED,
+            old_sigchld_handler: None,
+            wifsignaled: get_valid_wifsignaled(),
+        }
+    }
+}
+
+impl<'a, D, OT, H> ForkserverParent for UnicornAflForkserverParent<'a, D, OT, H>
+where
+    D: 'a,
+{
+    fn pre_fuzzing(&mut self) -> Result<(), libafl::Error> {
+        let old_sigchld_handler =
+            (unsafe { nix::sys::signal::signal(Signal::SIGCHLD, SigHandler::SigDfl) })
+                .inspect_err(|_| {
+                    error!("Fail to swap signal handler for SIGCHLD.");
+                })?;
+        self.old_sigchld_handler = Some(old_sigchld_handler);
+        Ok(())
+    }
+
+    fn handle_child_requests(&mut self) -> Result<i32, libafl::Error> {
+        let child_pipe_r = self.child_pipe_r.as_ref().unwrap().as_fd();
```

Comment:
Random question / potential future improvement: Unicorn doesn't really need forks, it could just run the emulator directly here and use snapshots, right?

Reply:
might be a lot faster and in-memory friendly, though. 
Instead of peristent iters, users should(tm) use mutli inputs (or custom mutators in afl++) and run the harness multiple times with each subsequent input, instead